### PR TITLE
Spark: don't shuffle row groups if training data requires non-shuffle

### PR DESCRIPTION
**Pull request by chongxiaoc**
In some cases (using pre-partitioned or sorted data), trainer doesn't
need data loader to shuffle rows. In this case, we should disable row groups
shuffling in Petastorm as well.

## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: horovod/spark/lightning/remote.py [hunk 1]
```diff
@@ -312,7 +312,7 @@ def calculate_shuffle_buffer_size():
         """
         import horovod.torch as hvd
 
-        if user_shuffle_buffer_size:
+        if user_shuffle_buffer_size != -1:
```
Comment (Tixxx) [inline horovod/spark/lightning/remote.py:315]: We should also return 0 if user_shuffle_buffer_size is zero or do zero-check in the caller.

File: horovod/spark/lightning/datamodule.py [hunk 2]
```diff
@@ -63,13 +63,16 @@ def setup(self, stage=None):
...
+                                               # Don't shuffle row groups without shuffling.
+                                               shuffle_row_groups=True if self.shuffle_size > 0 else False)
             if self.has_val:
                 self.val_reader = reader_factory(self.val_dir, num_epochs=self.num_reader_epochs,
                                                  cur_shard=self.cur_shard, shard_count=self.shard_count,
                                                  hdfs_driver=PETASTORM_HDFS_DRIVER,
                                                  schema_fields=self.schema_fields,
-                                                 storage_options=self.storage_options)
+                                                 storage_options=self.storage_options,
+                                                 shuffle_row_groups=False)
```
Comment (Tixxx) [inline horovod/spark/lightning/datamodule.py:75]: Good catch!

File: horovod/spark/lightning/remote.py (hunk 1 above)
Comment (EnricoMi) [reply horovod/spark/lightning/remote.py:315]: what if `user_shuffle_buffer_size < 0` here?

Comment (chongxiaoc) [reply horovod/spark/lightning/remote.py:315]: I think adding necessary zero-check in caller is more clear, also addresses valid value issue for < 0 case.
Will fix.

Comment (chongxiaoc) [reply horovod/spark/lightning/remote.py]: @Tixxx @EnricoMi fixed.

**Review by Tixxx — APPROVED** [review]

**Review by irasit — APPROVED** [review]

File: horovod/spark/keras/remote.py [hunk 3]
```diff
@@ -252,11 +256,13 @@ def train(serialized_model, train_rows, val_rows, avg_row_size):
                                     schema_fields=schema_fields,
                                     transform_spec=transform_spec,
                                     storage_options=storage_options,
+                                    shuffle_row_groups=False,
                                     **reader_factory_kwargs) \
                     if should_validate else empty_batch_reader() as val_reader:
 
                     train_data = make_dataset(train_reader, batch_size, shuffle_buffer_size,
-                                              is_batch_reader, shuffle=True, cache=inmemory_cache_all)
+                                              is_batch_reader, shuffle=True if shuffle_buffer_size > 0 else False,
```
[the remark this answers is not in the document]
Comment (chongxiaoc) [reply horovod/spark/keras/remote.py:264]: nope, this is the `shuffle` parameter tensorflow's dataloader takes.
`shuffle_row_groups` is for files, `shuffle` here is for data in memory.

File: horovod/spark/keras/remote.py [hunk 4]
```diff
@@ -241,6 +243,8 @@ def train(serialized_model, train_rows, val_rows, avg_row_size):
                                 schema_fields=schema_fields,
                                 transform_spec=transform_spec,
                                 storage_options=storage_options,
+                                # Don't shuffle row groups without shuffling.
```
[the remark this answers is not in the document]
Comment (chongxiaoc) [reply horovod/spark/keras/remote.py]: fixed

**Comment by chongxiaoc** [comment]
@EnricoMi  do you know if this Github Action `CI (Results) / Build and Test GPU heads (on Builtkite)` means a real failure or is just a flaky build?

File: horovod/spark/keras/remote.py [hunk 5]
```diff
@@ -112,10 +112,12 @@ def train(serialized_model, train_rows, val_rows, avg_row_size):
 
         pin_gpu(hvd, tf, k)
 
-        if not user_shuffle_buffer_size:
+        # If user specifies any user_shuffle_buffer_size (even 0), we should honor it.
+        if user_shuffle_buffer_size is None:
             shuffle_buffer_size = calculate_shuffle_buffer_size(
                 hvd, avg_row_size, train_rows / hvd.size())
         else:
+            assert user_shuffle_buffer_size >= 0, "user_shuffle_buffer_size cannot be negative!"
```
Comment (tgaddair) [inline horovod/spark/keras/remote.py:120]: Should prefer to raise a ValueError instead of asserting.

**Review by tgaddair — APPROVED** [review]

File: horovod/spark/keras/remote.py (hunk 5 above)
Comment (chongxiaoc) [reply horovod/spark/keras/remote.py:120]: Will do
